model/src/main/scala/aqua/model/topology/Topology.scala

```diff
@@ -96,10 +60,12 @@ object Topology {
 
         val prevOn = c.prevOnTags
 
-        val prevPath = prevOn.map { case OnTag(_, v) =>
-          v.reverse
+        // full paths from previous `on` tags
+        val prevPath = prevOn.map { case OnTag(c, v) =>
+          v.reverse :+ c
```

Comment:
Why `c` is needed there, what problem it solves? I suppose, if we have `via` paths, we might not need to get to the node itself. Do you have a case when that's another way?

Reply:
yes
```
func a(callback: -> ()):
  on node:
    callback()
    callback()
```
here we will ignore targets on path creation and the path will be
```
init_peer_id -> relay -> node -> relay -> init_peer_id -> relay -> init_peer_id
                                                                                           ^
```
without ignoring we could simplify `init_peer_id -> relay -> init_peer_id` to `init_peer_id`

Reply:
What do you mean by `target`?

Reply:
```
on `target` via `through1` via `through2`
   ....
```